clippy_lints/src/needless_traits_in_scope.rs

```diff
@@ -0,0 +1,76 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use rustc_errors::Applicability;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Find traits that are not explicitely used in scope (like `AsRef::as_ref()`)
+    /// but directly with the trait method (like `opt.as_ref()`).
+    ///
+    /// These traits can be imported anonymously with `use crate::Trait as _`.
+    /// This avoids name collision with other traits (possibly with the same name).
+    /// It also helps identify the traits in `use` statements.
+    ///
+    /// ### Why is this bad?
+    /// This needlessly brings a trait in trait's namespace.
```

Comment:
```suggestion
    /// This needlessly brings a trait in trait's namespace, where it could
    /// shadow other things. This is not really a problem, this lint is just
    /// for those who like to keep things tidy.
```

Reply:
Done in ef7fac20e.